src/storage/schema.rs

```diff
@@ -0,0 +1,265 @@
+use std::fmt;
+
+use crate::errors::*;
+use crate::storage::Transaction;
+
+pub const LATEST_SCHEMA_VERSION: SchemaVersion = SchemaVersion {
+    major: 0,
+    minor: 7,
+    patch: 0,
+    revision: 1,
+};
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SchemaVersion {
+    major: u32,
+    minor: u32,
+    patch: u32,
+    revision: u32,
+}
+
+impl SchemaVersion {
+    pub fn from_tuple(major: u32, minor: u32, patch: u32, revision: u32) -> Self {
+        Self {
+            major,
+            minor,
+            patch,
+            revision,
+        }
+    }
+}
+
+impl fmt::Display for SchemaVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}.{}.{}-{}",
+            self.major, self.minor, self.patch, self.revision
+        )
+    }
+}
+
+pub fn current_schema_version(tx: &mut Transaction) -> Result<Option<SchemaVersion>> {
+    let sql = "
+SELECT *
+FROM version";
+
+    fn convert_row(row: super::Row) -> Result<SchemaVersion> {
+        let revision = if row.column_count() > 3 {
+            row.get(3)?
+        } else {
+            // Might be prior to schema 0.7.0-1 where there was no revision column
+            0
+        };
+        Ok(SchemaVersion {
+            major: row.get(0)?,
+            minor: row.get(1)?,
+            patch: row.get(2)?,
+            revision,
+        })
+    }
+
+    let res = tx.query_single(sql, convert_row);
+
+    // Ignore errors
+    // TODO: this is similar to the Go implementation, but it would be better to ignore only the
+    // case where the table does not exist.
+    Ok(res.ok().flatten())
+}
+
+pub fn insert_schema_version(tx: &mut Transaction, version: &SchemaVersion) -> Result<()> {
+    let sql = "
+INSERT INTO version (major, minor, patch, revision)
+VALUES (?, ?, ?, ?)";
+
+    let params = rusqlite::params![
+        version.major,
+        version.minor,
+        version.patch,
+        version.revision
+    ];
+    match tx.execute_params(sql, params) {
+        // TODO: is all this error handling really necessary?
+        Ok(1) => Ok(()),
+        Ok(_) => {
+            Err("Version could not be inserted: expected exactly one row to be affected".into())
+        }
```

Comment:
This is the same as in Go, but I don't think this can really happen. It's not clear to me why we are extra paranoid here but not for other insertions/updates/deletions.